File: Java/01 - Numbers/03_Fibonacci_Sequence/Fibonacci.java
package com.example;

/*
    Project: Generate Fibonacci Sequence to nth number
    Author: Emma du Maurier
    Date: 16 April 2022
    Description: Enter a number and have the program generate the Fibonacci sequence up to that many places.

    Program Details: If the user enters 10 from the console input, then the following output will be displayed;
    0, 1, 1, 2, 3, 5, 8, 13, 21, 34
*/

import java.util.Scanner;

public class Main
{
    public static void main(String[] args)
    {
        Scanner scanner = new Scanner(System.in);
        int posOne = 0;
        int posTwo = 1;
        int f, end;

        do
        {
            System.out.println("How many numbers of the Fibonacci sequence to display?");
            System.out.println("Between 1 and 45 (inclusive)");
            end = scanner.nextInt();
            scanner.nextLine();
        }
        while ((end < 1) || (end > 45));

        // Check if the user entered '1' then display the first number of the sequence.
        if (end == 1)
        {
            System.out.println("Fibonacci Sequence:");
            System.out.println(posOne);
        }
        // Check if the user entered '2' then display the first and second numbers of the sequence.
        else if (end == 2)
        {
            System.out.println("Fibonacci Sequence:");
            System.out.println(posOne + ", " + posTwo);
        }
        // If the user entered 3 or higher, display the first two numbers of the sequence, then generate and display
        // each number after that until it reaches how many numbers the user entered.
        else
        {
            System.out.println("Fibonacci Sequence:");
            System.out.println(posOne + "\n" + posTwo);

            for (int i = 3; i <= end; i++)
            {
                f = posOne + posTwo;
                System.out.println(f);
                posOne = posTwo;
                posTwo = f;
            }
        }
    }
}

File: 01_Numbers/03_Fibonacci_Sequence/Fibonacci.cs
/*
 * Project: Generate Fibonacci Sequence to nth number
 * Author: Emma du Maurier
 * Date: 06 May 2024
 * Description: Enter a number and have the program generate the Fibonacci sequence up to that many places.
 * 
 * Program Details: If the user enters 10 from the console input, then the following output will be displayed;
 * 0, 1, 1, 2, 3, 5, 8, 13, 21, 34
*/

namespace Fibonacci
{
    internal class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Enter a number from 1 to 45 to generate the Fibonacci sequence up to that many places: ");
            int number = GetValidNumber();
            GenerateFibonacciSequence(number);
        }

        private static int GetValidNumber()
        {
            int number = 0;
            bool isValid = false;
            while (!isValid)
            {
                string input = Console.ReadLine();
                isValid = int.TryParse(input, out number) && number > 0;
                if (!isValid)
                {
                    Console.WriteLine("Invalid input. Please enter a positive integer: ");
                }
                else
                {
                    if (number > 45)
                    {
                        Console.WriteLine("Please enter a number that is 45 or under:");
                        isValid = false;
                    }
                }
            }
            return number;
        }

        private static void GenerateFibonacciSequence(int number)
        {
            int prevOne = 1, prevTwo = 0, fib;
            for (int i = 0; i < number; i++)
            {
                fib = prevOne + prevTwo;
                Console.WriteLine(fib);
                prevTwo = prevOne;
                prevOne = fib;
            }
        }
    }
}